### PR TITLE
Restrict dropping of specific extended stored procedures

### PR DESCRIPTION
**Pull request by roshan0708**
### Description

This commit adds a check to prevent dropping of specific extended stored procedures:
- xp_qv
- xp_instance_regread
- sp_addlinkedsrvlogin
- sp_droplinkedsrvlogin
- sp_dropserver
- sp_enum_oledb_providers
- sp_testlinkedserver

These procedures are critical and should not be dropped accidentally.

### Issues Resolved

Task: BABEL-4390

### Test Scenarios Covered ###
* **Use case based -** Yes

* **Boundary conditions -** Yes

* **Arbitrary inputs -** Yes

* **Negative test cases -** yes

* **Minor version upgrade tests -**

* **Major version upgrade tests -**

* **Performance tests -**

* **Tooling impact -**

* **Client tests -**

### Check List
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: contrib/babelfishpg_tsql/src/pl_handler.c [hunk 1]
```diff
@@ -6331,6 +6331,14 @@ bbf_ExecDropStmt(DropStmt *stmt)
 			if (schema_name != NULL)
 				logicalschema = get_logical_schema_name(schema_name, true);
 
+			/* Restrict dropping of specific procedures */
+			if (stmt->removeType == OBJECT_PROCEDURE && major_name && pg_strncasecmp(major_name, "xp_", 3) == 0)
```
Comment (shalinilohia50) [inline contrib/babelfishpg_tsql/src/pl_handler.c]: Can there be user defined procedures whose names start with "xp_*"? If yes, those should be allowed to drop.

Comment (roshan0708) [reply contrib/babelfishpg_tsql/src/pl_handler.c]: Right, there can be such cases. I'll update the logic for this as there are some more stored procedures which should be restricted from being droppped.

**Review by shalinilohia50 — APPROVED** [review]

**Comment by shalinilohia50** [comment]
1. Please update the PR description accordingly. Add `Signed-off-by:` too.
2. Shouldn't we add tests to other `schedule` files as well? Or just `latest/schedule` is enough?

File: contrib/babelfishpg_tsql/src/pl_handler.c [hunk 2]
```diff
@@ -6216,7 +6216,17 @@ bbf_ExecDropStmt(DropStmt *stmt)
...
-	const char *logicalschema = NULL;
+	const char		*logicalschema = NULL;
+	const char		*restricted_procedures[] = {
+		"xp_qv",
+		"xp_instance_regread",
+		"sp_addlinkedsrvlogin",
+		"sp_droplinkedsrvlogin",
+		"sp_dropserver",
+		"sp_enum_oledb_providers",
+		"sp_testlinkedserver"
```
Comment (thephantomthief) [inline contrib/babelfishpg_tsql/src/pl_handler.c:6233]: We should also restrict dropping any functions, views etc. that we are creating in master_dbo schema as part of Babelfish provisioning

File: contrib/babelfishpg_tsql/src/pl_handler.c [hunk 3]
```diff
@@ -6216,7 +6216,17 @@ bbf_ExecDropStmt(DropStmt *stmt)
 	Relation		relation = NULL;
 	Oid				schema_oid;
 	ListCell		*cell;
-	const char *logicalschema = NULL;
+	const char		*logicalschema = NULL;
+	const char		*restricted_procedures[] = {
```
Comment (thephantomthief) [inline contrib/babelfishpg_tsql/src/pl_handler.c:6226]: a comment explaining what restricted procedures mean would be good

**Review by thephantomthief — APPROVED** [review]

**Review by shardgupta — APPROVED** [review]

File: contrib/babelfishpg_tsql/src/pl_handler.c [hunk 4]
```diff
@@ -6216,7 +6216,23 @@ bbf_ExecDropStmt(DropStmt *stmt)
...
 	ListCell		*cell;
-	const char *logicalschema = NULL;
+	const char		*logicalschema = NULL;
+
+	/*
+	 * BABEL-4390
+	 * List of procedure names that are not allowed to be dropped. These procedures 
+	 * are considered essential or restricted due to security or operational reasons.
+	 */
+	const char		*restricted_procedures[] = {
```
Comment (shah-nirmit) [inline contrib/babelfishpg_tsql/src/pl_handler.c:6226]: Blocking: Should we not block dropping This procedures from PSQL Endpoint as well?

File: contrib/babelfishpg_tsql/src/pl_handler.c [hunk 5]
```diff
@@ -6331,6 +6347,22 @@ bbf_ExecDropStmt(DropStmt *stmt)
 			if (schema_name != NULL)
 				logicalschema = get_logical_schema_name(schema_name, true);
 
+			/* Restrict dropping of extended stored procedures */
+			if (stmt->removeType == OBJECT_PROCEDURE && major_name)
+			{
+				int i;
+				for (i = 0; i < RESTRICTED_PROCEDURES_COUNT; i++)
```
Comment (Deepesh125) [inline contrib/babelfishpg_tsql/src/pl_handler.c:6354]: We should also check the schema, just procedure name comparison is not enough. End user may have procedure defined with the same name under user defined schema. Current logic does not handle it well.

**Pull request merged**